### PR TITLE
[SPARK-32052][SQL] Extract common code from date-time field expressions

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
Extract common code from the expressions that get date or time fields from input dates/timestamps to new expressions `GetDateField` and `GetTimeField`, and re-use the common traits from the affected classes.

### Why are the changes needed?
Code deduplication improves maintainability. 

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
By `DateExpressionsSuite`

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 1]
```diff
@@ -274,39 +274,43 @@ case class DateSub(startDate: Expression, days: Expression)
...
-    Examples:
-      > SELECT _FUNC_('2009-07-30 12:58:59');
-       12
-  """,
-  group = "datetime_funcs",
-  since = "1.5.0")
-case class Hour(child: Expression, timeZoneId: Option[String] = None)
-  extends UnaryExpression with TimeZoneAwareExpression with ImplicitCastInputTypes
-    with NullIntolerant {
+trait GetTimeField extends UnaryExpression
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:277]: When I made it as `abstract class` with 2 args `func` and `funcName`, I got the following exception on tests from `DateExpressionsSuite`:
```

```

**Comment by MaxGekk** [comment]
@cloud-fan @srowen @dongjoon-hyun @HyukjinKwon Could you review the refactoring, please.

**Review by Fokko — APPROVED** [review]
Nice one

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by HyukjinKwon** [comment]
late LGTM too

**Pull request closed without merging**